R/qenv-replace_code.R

```diff
@@ -32,19 +32,24 @@ setGeneric("replace_code", function(object, code) standardGeneric("replace_code"
 #' @keywords internal
 setMethod("replace_code", signature = c("qenv", "character"), function(object, code) {
   masked_code <- get_code(object)
-  masked_code[length(masked_code)] <- code
-  object@code <- masked_code
+  code_lines <- unlist(strsplit(masked_code, "\n"))
+
+  if (!is.null(code_lines)) {
+    code_lines[length(code_lines)] <- code
+    object@code <- paste(code_lines, collapse = "\n")
+  }
+
   object
 })
 
 #' @keywords internal
 setMethod("replace_code", signature = c("qenv", "language"), function(object, code) {
-  replace_code(object, code = format_expression(code))
+  replace_code(object, code = paste(lang2calls(code), collapse = "\n"))
 })
 
 #' @keywords internal
 setMethod("replace_code", signature = c("qenv", "expression"), function(object, code) {
-  replace_code(object, code = format_expression(code))
+  replace_code(object, code = paste(lang2calls(code), collapse = "\n"))
```

Comment:
I'm having second thoughts about removing `format_expression`. It's used in multiple places. I think it can stay.

Reply:
I had a similar thought previously, but following a conversation with @chlebowa it was clarified that the function simply uses paste with a designated collapse parameter after processing with `lang2calls`. The `format_expression` function originally had a more complex design but it was simplified into a simple one-liner in a past refactoring phase. Currently, it's essentially redundant.

Reply:
format_expression sounds like a wrapper for paste, where you overwrite default collapse param :)

Reply:
Did you guys consider having `lang2calls` return the `paste(...)`?

It seems that it is only used this way in the package _(other than in tests)_

ps. and rename it to `lang2str` or `lang2character`?

Reply:
there is recursion in `lang2calls`  and  implementing the paste function with collapse directly inside `lang2calls` can lead to unintended results due to concatenation of results from each recursive call.